rust-src/bulletproofs/src/utils.rs

```diff
@@ -0,0 +1,139 @@
+//! Shared functions used by the proofs in this crate
+use crypto_common::*;
+use crypto_common_derive::*;
+use curve_arithmetic::Curve;
+use ff::Field;
+#[cfg(test)]
+use rand::Rng;
+/// Struct containing generators G and H needed for range proofs
+#[allow(non_snake_case)]
+#[derive(Debug, Clone, Serialize, SerdeBase16Serialize)]
+pub struct Generators<C: Curve> {
+    #[size_length = 4]
+    pub G_H: Vec<(C, C)>,
+}
+
+impl<C: Curve> Generators<C> {
+    /// **Warning** do not use in production!
+    /// This generates a list of generators of a given size for
+    /// testing purposes. For production, generators must be created such that
+    /// discrete logarithms between different generators are not known, which is
+    /// not guaranteed by this function.
+    #[cfg(test)]
+    pub(crate) fn generate(n: usize, csprng: &mut impl Rng) -> Self {
+        let mut gh = Vec::with_capacity(n);
+        for _ in 0..n {
+            let x = C::generate(csprng);
+            let y = C::generate(csprng);
+            gh.push((x, y));
+        }
+        Self { G_H: gh }
+    }
+
+    /// Returns the prefix of length nm of a given generator.
+    /// This function panics if nm > length of the generator.
+    pub fn take(&self, nm: usize) -> Self {
+        Self {
+            G_H: self.G_H[0..nm].to_vec(),
+        }
+    }
+}
+
+/// This function takes one argument n and returns the
+/// vector (z^j, z^{j+1}, ..., z^{j+n-1}) in F^n for any field F
+/// The arguments are
+/// - z - the field element z
+/// - first_power - the first power j
+/// - n - the integer n.
+pub fn z_vec<F: Field>(z: F, first_power: usize, n: usize) -> Vec<F> {
+    let mut z_n = Vec::with_capacity(n);
+    // let mut z_i = F::one();
+    // FIXME: This should would be better to do with `pow`.
+    // for _ in 0..first_power {
+    //    z_i.mul_assign(&z);
+    //}
+    let exp: [u64; 1] = [first_power as u64];
+    let mut z_i = z.pow(exp);
+    for _ in 0..n {
+        z_n.push(z_i);
+        z_i.mul_assign(&z);
+    }
+    z_n
+}
+
+/// Converts the u64 set vector into a vector over the field
+pub fn get_set_vector<C: Curve>(the_set: &[u64]) -> Vec<C::Scalar> {
+    let n = the_set.len();
+    let mut s_vec = Vec::with_capacity(n);
+    for elem_i in the_set {
+        let s_i = C::scalar_from_u64(*elem_i);
+        s_vec.push(s_i);
+    }
+    s_vec
+}
+
+/// Pads a non-empty field vector to a power of two length by repeating the last
+/// element For empty vectors the function is the identity.
+pub fn pad_vector_to_power_of_two<F: Field>(vec: &mut Vec<F>) {
+    let n = vec.len();
+    if n == 0 {
+        return;
+    }
+    let k = n.next_power_of_two();
+    if let Some(last) = vec.last().cloned() {
+        let d = k - n;
+        for _ in 0..d {
+            vec.push(last)
+        }
+    }
```

Comment:
This function is not used apart from in the test that it works correctly.

Thus it should generally not be part of the PR. Why is it here?

Reply:
I is probably needed at some higher level. We may want to prove statements about sets of size not a power of 2. And there it is crucial to extend the set properly, i.e., changing the multiset without changing the set.

Reply:
We could remove it here, but it will be needed if we want to prove statements about sets with arbitrary not a power of 2. Should we create adapter functions that allow to prove statements for sets of arbitrary size?

Reply:
See my comment below. You have a good point that this function should stay, and we should probably just use it in the prover and verifier (at least the ones we expose).

Reply:
Now used in prove and verify.